core/matrix/ell.cpp

```diff
@@ -0,0 +1,177 @@
+#include <vector>
+#include <iostream>
+
+namespace gko {
+namespace matrix {
+
+
+namespace {
+
+
+template <typename... TplArgs>
+struct TemplatedOperation {
+    // GKO_REGISTER_OPERATION(spmv, ell::spmv<TplArgs...>);
+    // GKO_REGISTER_OPERATION(advanced_spmv, ell::advanced_spmv<TplArgs...>);
+    // GKO_REGISTER_OPERATION(convert_to_csr, ell::convert_to_dense<TplArgs...>);
+    // GKO_REGISTER_OPERATION(move_to_csr, ell::move_to_dense<TplArgs...>);
+};
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::apply(const LinOp *b, LinOp *x) const
+{
+    NOT_IMPLEMENTED;
+    // ASSERT_CONFORMANT(this, b);
+    // ASSERT_EQUAL_ROWS(this, x);
+    // ASSERT_EQUAL_COLS(b, x);
+    // using Dense = Dense<ValueType>;
+    // this->get_executor()->run(
+    //     TemplatedOperation<ValueType, IndexType>::make_spmv_operation(
+    //         this, as<Dense>(b), as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::apply(const LinOp *alpha, const LinOp *b,
+                                      const LinOp *beta, LinOp *x) const
+{
+    NOT_IMPLEMENTED;
+    // ASSERT_CONFORMANT(this, b);
+    // ASSERT_EQUAL_ROWS(this, x);
+    // ASSERT_EQUAL_COLS(b, x);
+    // ASSERT_EQUAL_DIMENSIONS(alpha, size(1, 1));
+    // ASSERT_EQUAL_DIMENSIONS(beta, size(1, 1));
+    // using Dense = Dense<ValueType>;
+    // this->get_executor()->run(
+    //     TemplatedOperation<ValueType, IndexType>::make_advanced_spmv_operation(
+    //         as<Dense>(alpha), this, as<Dense>(b), as<Dense>(beta),
+    //         as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::convert_to(Dense<ValueType> *other) const
+{
+    NOT_IMPLEMENTED;
+    // other->set_dimensions(this);
+    // other->values_ = values_;
+    // other->col_idxs_ = col_idxs_;
+    // other->max_nnz_row_ = max_nnz_row_;
+}
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::move_to(Dense<ValueType> *other)
+{
+    NOT_IMPLEMENTED;
+    // other->set_dimensions(this);
+    // other->values_ = std::move(values_);
+    // other->col_idxs_ = std::move(col_idxs_);
+    // other->max_nnz_row_ = std::move(max_nnz_row_);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::read_from_mtx(const std::string &filename)
+{
+    auto data = read_raw_from_mtx<ValueType, IndexType>(filename);
+
+    // Count number of nonzero elements of every row.
+    size_type nnz = 0;
+    std::vector<index_type> nnz_row(data.num_rows, 0);
+    for (const auto &elem : data.nonzeros) {
+        nnz += (std::get<2>(elem) != zero<ValueType>());
+        nnz_row.at(std::get<0>(elem))++;
+    }
+
+    // Get number of maximum nonzeros in a single row.
+    index_type max_nnz_row = 0;
+    for (const auto &elem : nnz_row) {
+        max_nnz_row = std::max(max_nnz_row, elem);
+    }
```

Comment:
@gflegar, where would you suggest to put this function in?

Reply:
Just define it in this source file for now, it shouldn't appear in the headers. Make sure you wrap it into an anonymous namespace to prevent the symbol being exported from the compilation unit.